include/exadg/grid/grid_data.h

```diff
@@ -67,6 +71,16 @@ struct GridData
     print_parameter(pcout, "Mapping degree", mapping_degree);
   }
 
+  void
+  add_parameters(dealii::ParameterHandler & prm, std::string const & subsection_name = "Grid")
+  {
+    // clang-format off
+    prm.enter_subsection(subsection_name);
+      prm.add_parameter("GridFile", grid_file, "External input grid file.");
+    prm.leave_subsection();
+    // clang-format on
+  }
```

Comment:
This design is not in line with what we currently do in ExaDG. Note that `OutputParameters` (depending on `dealii::ParameterHandler`) and `OutputDataBase` are two different classes. So the philosophy behind this design is that `dealii::ParameterHandler` should not be visible in "inner" data structures, e.g. `GridData`. When following the choice in this PR, we give up the flexbility to use something different than `dealii::ParameterHandler` (there are already other variants in deal.II itself). I currently see it this way: `Application` depends on `ParameterHandler` and `Grid`/`GridData`, but  `Grid`/`GridData` should not depend on `Application` and `ParameterHandler`.

I am not sure which design is the best one, but we definitely need to discuss this.

Reply:
Yes, this was the initial reason why I added `InputParameters`. Because it felt natural after seeing `OutputParameters` and I forgot the existence of `gridfile` in `gridData` :) 

I don't see it as a complete dependency, i.e., one could still use `gridData` without using the `add_parameters` function. If one wants to read an external grid, yes we are dependent, which makes sense.

What I could do to solve this is introducing `GridParameters` with a free function `add_parameters` with a write access to `GridData`? Or a free function in `GridUtilities`?

Reply:
In any case, I think one has the dependency software-wise.

I think I am against a write access to `GridData`. Maybe we wait some time with this aspect for other opinions before doing too many changes. I do not say that the current implementation has to be changed, but want to point to weaknesses from a design perspective.

Reply:
Only this is not addressed in the latest commits. I couldn't find a solution, where we:

-  don't allow write access to `GridData`
-  don't introduce new dependencies  
-  and keep `grid_file` string in `GridData`.

Reply:
`OutputParameters` and `OutputDataBase` is one such realization, as mentioned above.

I am not sure what you mean by write access exactly. All these parameter structs have public members, implying write access. I am against something like `grid_data.set_filename(grid_parameters)`, since this introduces not only additional or unnecessary member functions, but also additional dependencies. However, an assignment `grid_data.filename = name` (just as we set all the parameters in application.h) is fine, isn't it? This has to happen in `Application::set_parameters()`, because from then on, `grid_data` needs to be considered `const`.

Reply:
> However, an assignment `grid_data.filename = name` (just as we set all the parameters in application.h) is fine, isn't it?

This I would have described as "write access", since we are writing on a parameter, and I thought you were against it. I will change the design to the suggested version!

Reply:
Wait a second. You do not have to change it to that version. I mainly want to have a discussion about this. I understand that you want to finish the present PR soon, so we can also take the current version and change it in the future.

Reply:
Regarding "write access": we make the member variables of `Parameters` public so that we do not always have to call setter/getter functions e.g. in `Application::set_parameters()`. At the same time, we do not want to misuse this "freedom" in the sense that we pass the structs to functions that can manipulate the parameters. Imagine I pass `gird_data` to two different functions during the setup phase, both having write access to `grid_data`. How do you know that the data in `grid_data` is reasonable after setup?

Reply:
> Wait a second. You do not have to change it to that version. I mainly want to have a discussion about this. I understand that you want to finish the present PR soon, so we can also take the current version and change it in the future.

Maybe it seemed like I wanted to finish this PR soon and that is why I accept your suggestion, but I am trying to contribute a good PR, not something to be changed in the future. That is why I am listening to suggestions and making changes accordingly.  

I liked the idea of having a separate file `grid_parameters.h` with `add_parameters` in it. I know I don't have to change it to that version.

Reply:
I am not completely whether I understood the discussion here. The discussion is whether to have a method called `add_parameters()` or to have a free function for each parameter struct similar to what we have for the enums. Personally, I normally use the first approach in my projects but the free-function approach is also fine if that is more inline with rest of ExaDG :+1:

Reply:
I guess we had a misunderstanding with @nfehn, and a short meeting solved it :+1:  We thought adding `grid_parameters.h` with `add_parameters()` in it would be a good approach here, also considering your comment @peterrum.

I will implement this solution!

Reply:
I have made the changes in 3f0998180eafacf5a49545f48f19b5f3e82f6f04.